viogpu/common/viogpu_idr.cpp

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright (C) 2019-2020 Red Hat, Inc.
+ *
+ * Written By: Vadim Rozenfeld <vrozenfe@redhat.com>
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met :
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and / or other materials provided with the distribution.
+ * 3. Neither the names of the copyright holders nor the names of their contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include "helper.h"
+#include "baseobj.h"
+
+VioGpuIdr::VioGpuIdr()
+{
+    m_uStartIndex = 0;
+    m_IdBitMap.SizeOfBitMap = 0;
+    m_IdBitMap.Buffer = NULL;
+}
+
+VioGpuIdr::~VioGpuIdr()
+{
+    Close();
+}
+
+BOOLEAN VioGpuIdr::Init(
+    _In_ ULONG start)
+{
+    PUCHAR buf = NULL;
+    BOOLEAN ret = FALSE;
+    m_uStartIndex = start;
+    ExInitializeFastMutex(&m_IdBitMapMutex);
+
+    VIOGPU_ASSERT(m_IdBitMap.Buffer == NULL);
+    Lock();
+    buf = new (NonPagedPoolNx) UCHAR[PAGE_SIZE];
+    if (buf) {
+        RtlInitializeBitMap(&m_IdBitMap,
+            (PULONG)buf,
+            CHAR_BIT * PAGE_SIZE);
+        RtlClearAllBits(&m_IdBitMap);
+        RtlSetBits(&m_IdBitMap, 0, m_uStartIndex);
+        ret = TRUE;
+    }
+    Unlock();
+    return ret;
+}
+
+#pragma warning( disable: 28167 )
+//_IRQL_raises_(APC_LEVEL)
+//_IRQL_saves_global_(OldIrql, m_IdBitMapMutex)
+VOID VioGpuIdr::Lock(VOID)
```

Comment:
Yep, I aware of scoped locking pattern and used it many times in the past. There will be lot of C++ related changes in the future. But for now I would prefer pushing the code upstream and backporting it to our build system rather than polishing the code that is going to be changed in any case when we start working on Full WDDM driver. 
Best,
Vadim.